xml/src/main/scala/org/broadinstitute/monster/extractors/xml/XmlExtractor.scala

```diff
@@ -0,0 +1,108 @@
+  writeJson: (File, XmlExtractor.GcsObject) => IO[Unit]
+)(implicit context: ContextShift[IO]) {
+  import XmlExtractor.GcsObject
+
+  def extract(
+    input: GcsObject,
+    output: GcsObject,
+    xmlTag: String,
+    tagsPerFile: Int
+  ): IO[Unit] = {
+    getXml(input).through(xmlToJson(xmlTag, tagsPerFile)).evalMap { jsonFile =>
+      IO.pure(jsonFile).bracket(writeJson(_, output)) { file =>
+        IO.delay(file.delete())
+      }
+    }
+  }.compile.drain
+
+  private def xmlToJson(xmlTag: String, tagsPerFile: Int): Pipe[IO, Byte, File] = { xml =>
+    val jsonXMLConfig = new JsonXMLConfigBuilder()
+      .autoArray(true)
+      .autoPrimitive(true)
+      .repairingNamespaces(true)
+      .virtualRoot(xmlTag)
+      .build()
+
+    def groupXmlTags(
+      xmlEventStream: Stream[IO, XMLEvent],
+      eventsAccumulated: Chain[XMLEvent],
+      inTag: Boolean
+    ): Pull[IO, Chain[XMLEvent], Unit] = {
+      xmlEventStream.pull.uncons1.flatMap {
+        case None => Pull.done
+        case Some((xmlEvent, remainingXmlEvents)) =>
+          if (xmlEvent.isEndElement && xmlEvent
+                .asEndElement()
+                .getName
+                .getLocalPart == xmlTag) {
+            Pull.output1(eventsAccumulated.append(xmlEvent)).flatMap { _ =>
+              groupXmlTags(remainingXmlEvents, Chain.empty, inTag = false)
+            }
+          } else if (inTag || xmlEvent.isStartElement && xmlEvent
+                       .asStartElement()
+                       .getName
+                       .getLocalPart == xmlTag) {
+            groupXmlTags(xmlEventStream, eventsAccumulated.append(xmlEvent), inTag = true)
+          } else {
+            groupXmlTags(remainingXmlEvents, eventsAccumulated, inTag)
+          }
+      }
+    }
+
+    val xmlEventStream = xml.through(fs2.io.toInputStream).flatMap { inputStream =>
+      val reader = newInstance().createXMLEventReader(inputStream)
+      Stream.fromIterator[IO, XMLEvent](new Iterator[XMLEvent] {
+        override def hasNext: Boolean = reader.hasNext
+        override def next(): XMLEvent = reader.nextEvent()
+      })
+    }
+
+    groupXmlTags(xmlEventStream, Chain.empty, inTag = false).stream
+      .chunkN(tagsPerFile)
+      .evalMap { xmlChunk =>
+        IO.delay(File.newTemporaryFile()).flatMap { file =>
+          IO.delay(file.newOutputStream)
+            .bracket { outputStream =>
+              xmlChunk.traverse { xmlEvents =>
+                IO.delay {
+                  new JsonXMLOutputFactory(jsonXMLConfig)
+                    .createXMLEventWriter(outputStream)
+                }.bracket { writer =>
+                  xmlEvents.traverse_ { xmlEvent =>
+                    IO.delay {
+                      writer.add(xmlEvent)
+                    }
+                  }
+                } { writer =>
+                  IO.delay {
+                    writer.close()
+                    outputStream.write('\n')
```

Comment:
I would move this `write` call out of the teardown method in the `bracket` / `Resource`, since it doesn't have anything to do with cleaning up the XML writer. You can chain it on to the `bracket` instead.